File: problems/Codeforces/1791D/solution.cpp
#include <algorithm>
#include <iostream>
#include <string>
#include <vector>

int main() {
    using namespace std;
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int t;
    cin >> t;
    while(t--) {
        int _;
        string s;
        cin >> _ >> s;
        vector<int> left_freq(26, 0);
        vector<int> right_freq(26, 0);
        uint32_t left_distinct = 0;
        uint32_t right_distinct = 0;
        uint32_t max_cnt = 0;
        for(char& c: s) {
            int idx = static_cast<int>(c - 'a');
            if (right_freq[idx]++ == 0) {
                right_distinct++;
            }
        }

        for(char& c: s) {
            int idx = static_cast<int>(c - 'a');
            if(left_freq[idx]++ == 0) {
                left_distinct++;
            }
            if(--right_freq[idx] == 0) {
                right_distinct--;
            }
            max_cnt = max(max_cnt, left_distinct+right_distinct);
        }
        cout << max_cnt << '\n';
    }
    return 0;
}

File: problems/codeforces/1791D/solution.rs
use std::io::{self, Read, Write, BufWriter};

fn main() -> io::Result<()> {
    let mut input = String::new();
    io::stdin().lock().read_to_string(&mut input)?;
    let mut iter = input.split_ascii_whitespace();
    let mut writer = BufWriter::new(io::stdout().lock());
    let t: usize = iter.next().unwrap().parse().unwrap();
    for _ in 0..t {
        iter.next();
        let s = iter.next().unwrap().as_bytes();
        let mut left_freq = [0; 26];
        let mut right_freq = [0; 26];
        let mut left_distinct = 0;
        let mut right_distinct = 0;
        let mut max_cnt = 0;
        for &c in s {
            let idx = (c - b'a') as usize;
            if right_freq[idx] == 0 {
                right_distinct += 1;
            }
            right_freq[idx] += 1;
        }

        for &c in s {
            let idx = (c - b'a') as usize;
            if left_freq[idx] == 0 {
                left_distinct += 1;
            }
            left_freq[idx] += 1;

            right_freq[idx] -= 1;
            if right_freq[idx] == 0 {
                right_distinct -= 1;
            }

            max_cnt = max_cnt.max(left_distinct + right_distinct);
        }
        writeln!(writer, "{}", max_cnt)?;
    }
    Ok(())
}
